### PR TITLE
Add Dart support

### LINKED ISSUE #3099: [COMPILER REQUEST]: Dart, the language with AOT used by Flutter, which is a popular cross-platform rapid-development mobile framework. With compiler-explorer, we can quickly understand performance - since mobile apps are performance-critical
### Compiler name

Dart

### Compiler version

latest

### Compiler language

Dart

### Compiler homepage

https://dart.dev/overview#native-platform or https://dart.dev/

### Motivation

Dart is the language used by Flutter. Flutter is a popular cross-platform rapid-development mobile framework, and many many many people are using it to develop apps. (For example, I have written >100kloc of Dart code to develop my Flutter app).

As we know, performance is very critical for mobile apps, since phones usually do not have a powerful CPU and users never like an app that runs slowly. Thus, with godbolt, we can easily gain insights for how Dart code is compiled by the Dart AOT compiler into native machine language.

Remark: Dart has both AOT and JIT. Here I am proposing that the AOT compiler to be integrated into godbolt.

### PR DESCRIPTION
**Pull request by miDeb**
Closes #3099

This adds support for running the dart compiler (it was already installed since https://github.com/compiler-explorer/infra/pull/627). For now this means it supports dart version 2.14.4 (newest version would be 2.16). I hope those changes make sense, but it's likely I've missed something :)

### CODE REVIEW

File: lib/asm-parser.js [hunk 1]
```diff
@@ -591,6 +591,7 @@ export class AsmParser extends AsmRegex {
             if (match) {
                 maybeRemovePreviousLabel();
                 mayRemovePreviousLabel = filters.libraryCode;
+                source = null;
```
Comment (partouf) [inline lib/asm-parser.js:594]: This is not correct for most compilers and assembly. Perhaps overriding this class and function would be better than to cause bugs with this.

File: lib/asm-parser.js [hunk 2]
```diff
@@ -600,17 +612,8 @@ export class AsmParser extends AsmRegex {
...
-                    const lastLine = asm[asm.length - 1];
-                    if (lastLine.text && this.labelDef.test(lastLine.text)) {
-                        asm.pop();
-                    }
-                    mayRemovePreviousLabel = false;
-                }
-                continue;
-            } else {
-                mayRemovePreviousLabel = true;
+            if (isMainsource) {
```
Comment (partouf) [inline lib/asm-parser.js:615]: this also looks like a breaking change

**Review by partouf — CHANGES_REQUESTED** [review]
Thanks for doing this!
I think you'll have to override the parser functions in `asm-parser-dart` instead of changing `asm-parser`, this breaks too much as far as I can see

**Comment by miDeb** [comment]
Thanks for the quick review! I moved my changes to `asm-parser-dart.js` and removed them from `asm-parser.js`.

**Review by partouf — APPROVED** [review]
Thanks for the quick changes. Looks good to me

**Comment by partouf** [comment]
I have disabled execution for now, it looks like it didn't actually create an executable (and didn't have +x), probably just an object file

```
$ readelf -h output.s
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - System V
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x0
  Start of program headers:          64 (bytes into file)
  Start of section headers:          894448 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         6
  Size of section headers:           64 (bytes)
  Number of section headers:         15
  Section header string table index: 9
```

**Comment by miDeb** [comment]
Sorry, I forgot to copy the execution wrapper to the actual config file. It created an "aot snapshot" which should be executed with `dartaotruntime`.

**Comment by miDeb** [comment]
I opened #3357 to (hopefully) fix it.

**Comment by fzyzcjy** [comment]
Wow thanks for handling dart support!

**Pull request merged**